src/dmd/typinf.d

```diff
@@ -29,9 +29,16 @@ import dmd.visitor;
 extern (C++) void genTypeInfo(Type torig, Scope* sc)
 {
     //printf("Type::genTypeInfo() %p, %s\n", this, toChars());
+
+    if (!global.params.useTypeInfo)
+    {
+        torig.error(Loc.initial, "`TypeInfo` cannot be used with -betterC");
+        fatal();
+    }
```

Comment:
This block of code was modified to make it consistent with the following in statementsem.d

https://github.com/dlang/dmd/blob/ec4534ec63958b17167d636cd8b98824ce52f7d3/src/dmd/statementsem.d#L3626-L3636

Reply:
Can we have a valid line for the error?

Reply:
Does this error apply for both explicit and implicit uses of RTTI?

Reply:
> Does this error apply for both explicit and implicit uses of RTTI?

It applies to anything that ends up calling `genTypeInfo`, which, under the current implementation could be implicit or explicit uses of `TypeInfo`.  For example, trying to do a string concatenation will eventually call runtime hooks like `_d_arraycatnTX` which takes `TypeInfo` as one of its arguments.  See https://issues.dlang.org/show_bug.cgi?id=18312

IMO all of those runtime hooks should be changed to templates, that don't require `TypeInfo`, so we shouldn't need `TypeInfo` for string concatenation as long as the runtime hooks are implemented.  That's something I hope to champion after this PR (maybe 1 or 2 more) are merged.

TL;DR: Yes implicit and explicit uses of `TypeInfo` will trigger the error, but by replacing runtime hooks with templates (later) we may be able to eliminate such errors.

Reply:
Ok, but I'd still feel better if explicit uses got a file/line location to highlight where you're using RTTI.

Reply:
By "explicit uses" are you referring to `typeid`?  That's the only case I can think of where `TypeInfo` is explicitly used, and even then, it's indirect.  Please let me know if you have something else in mind.

Reply:
I think showing the location of implicit uses is even more desirable, as you cannot grep for these as with explicit uses of `typeid`. That would mean to pass the location of an expression via semanticTypeInfo and getTypeInfoType.

That will probably also be helpful when generating type info more lazily as in https://github.com/dlang/dmd/pull/6561

Reply:
OK, I'll explore this.  Marking this as WIP while I try.